File: benchmarks/python/bench.py
"""
FFVM Cross-Language Benchmark — Python
Implements B01–B05 with IDENTICAL logic to FFVM/C# benchmarks.
Each benchmark uses int for loop control and float for computation.
B02 is pure integer (classic Fibonacci).

Usage: python3 bench.py
Output: [XLANG] name | lang | us | scale | result
"""
import time


def measure(name: str, fn, scale: int, expected) -> None:
    # warmup
    for _ in range(20):
        fn(scale)

    # measure
    runs = 200
    t0 = time.perf_counter()
    result = 0
    for _ in range(runs):
        result = fn(scale)
    t1 = time.perf_counter()
    us = ((t1 - t0) / runs) * 1e6

    status = "PASS" if result == expected else f"FAIL(got={result},want={expected})"
    print(f"[XLANG] {name} | python | {us:.1f} | {scale} | {result} | {status}")


# B01: ArithLoop — int loop, float arithmetic
def b01_arith_loop(n: int) -> float:
    acc = 0.0
    for i in range(n):
        x = i + 0.5
        acc += x
        temp = x * 2.0
        temp -= 1.0
        acc += temp
    return acc


# B02: Fibonacci — pure int
def b02_fibonacci(n: int) -> int:
    a, b = 0, 1
    for _ in range(n):
        a, b = b, a + b
    return a


# B03: NestedLoop — int loops, float multiply-accumulate
def b03_nested_loop(n: int) -> float:
    acc = 0.0
    i = 0
    while i < n:
        j = 0
        while j < n:
            acc += (i + 0.5) * (j + 0.5)
            j += 1
        i += 1
    return acc


# B04: Branching — int loop+branch, float accumulate
def b04_branching(n: int) -> float:
    acc = 0.0
    i = 0
    while i < n:
        x = i * 0.5
        m = i % 4
        if m == 0:
            acc += x
        elif m == 1:
            acc += x * 2.0
        elif m == 2:
            acc += x * 0.5
        else:
            acc += x * 4.0
        i += 1
    return acc


# B05: Accumulator — int loop, float sum
def b05_accumulator(n: int) -> float:
    s = 0.0
    i = 0
    while i < n:
        s += i * 0.5
        i += 1
    return s


if __name__ == "__main__":
    print("[XLANG_START] python")
    measure("B01_ArithLoop",   b01_arith_loop,  10000, b01_arith_loop(10000))
    measure("B02_Fibonacci",   b02_fibonacci,   46,    b02_fibonacci(46))
    measure("B03_NestedLoop",  b03_nested_loop, 100,   b03_nested_loop(100))
    measure("B04_Branching",   b04_branching,   10000, b04_branching(10000))
    measure("B05_Accumulator", b05_accumulator, 50000, b05_accumulator(50000))
    print("[XLANG_END] python")

File: benchmarks/lua/bench.lua
-- ============================================================
--  FFVM Cross-Language Benchmark — Lua
--  Implements B01–B05 with IDENTICAL logic to FFVM/C# benchmarks.
--  Each benchmark uses int for loop control and float for computation.
--  B02 is pure integer (classic Fibonacci).
--  Lua 5.1 has only one number type (double) — degraded representation.
--
--  Usage: lua bench.lua
--  Output: [XLANG] name | lang | us | scale | result
-- ============================================================

local clock = os.clock

local function measure(name, fn, scale, expected)
    -- warmup
    for _ = 1, 20 do fn(scale) end

    -- measure
    local runs = 200
    local t0 = clock()
    local result
    for _ = 1, runs do
        result = fn(scale)
    end
    local t1 = clock()
    local us = ((t1 - t0) / runs) * 1e6

    local status = (result == expected) and "PASS" or
        string.format("FAIL(got=%s,want=%s)", tostring(result), tostring(expected))
    print(string.format("[XLANG] %s | lua | %.1f | %d | %s | %s",
        name, us, scale, tostring(result), status))
end

-- B01: ArithLoop — int loop, float arithmetic
local function b01(n)
    local acc = 0.0
    for i = 0, n - 1 do
        local x = i + 0.5
        acc = acc + x
        local temp = x * 2.0
        temp = temp - 1.0
        acc = acc + temp
    end
    return acc
end

-- B02: Fibonacci — pure int, iterative fib(N)
local function b02(n)
    local a, b = 0, 1
    for _ = 1, n do
        a, b = b, a + b
    end
    return a
end

-- B03: NestedLoop — int loops, float multiply-accumulate
local function b03(n)
    local acc = 0.0
    for i = 0, n - 1 do
        for j = 0, n - 1 do
            acc = acc + (i + 0.5) * (j + 0.5)
        end
    end
    return acc
end

-- B04: Branching — int loop+branch, float accumulate
local function b04(n)
    local acc = 0.0
    for i = 0, n - 1 do
        local x = i * 0.5
        local m = i % 4
        if m == 0 then acc = acc + x
        elseif m == 1 then acc = acc + x * 2.0
        elseif m == 2 then acc = acc + x * 0.5
        else acc = acc + x * 4.0 end
    end
    return acc
end

-- B05: Accumulator — int loop, float sum
local function b05(n)
    local sum = 0.0
    for i = 0, n - 1 do
        sum = sum + i * 0.5
    end
    return sum
end

-- Expected results — iterative for exact match
local function b01_expected(n) return b01(n) end
local function b02_expected(n)
    local a, b = 0, 1
    for _ = 1, n do a, b = b, a + b end
    return a
end
local function b03_expected(n) return b03(n) end
local function b04_expected(n) return b04(n) end
local function b05_expected(n) return b05(n) end

print("[XLANG_START] lua")
measure("B01_ArithLoop",   b01, 10000, b01_expected(10000))
measure("B02_Fibonacci",   b02, 46,    b02_expected(46))
measure("B03_NestedLoop",  b03, 100,   b03_expected(100))
measure("B04_Branching",   b04, 10000, b04_expected(10000))
measure("B05_Accumulator", b05, 50000, b05_expected(50000))
print("[XLANG_END] lua")
